File: src/shared/models/timeline.py
"""
时间线数据结构

用于视频剪辑的时间线、轨道、片段数据模型
"""
from typing import List, Dict, Any, Optional
from dataclasses import dataclass, field
from datetime import datetime
import json


@dataclass
class TimelineClip:
    """时间线片段"""
    id: str
    material_id: int
    material_name: str
    start: float  # 在时间线上的开始时间（秒）
    end: float    # 在时间线上的结束时间（秒）
    trim_start: float = 0.0  # 素材裁剪开始点
    trim_end: float = 0.0    # 素材裁剪结束点
    speed: float = 1.0       # 播放速度
    volume: float = 1.0      # 音量
    effects: List[Dict] = field(default_factory=list)

    @property
    def duration(self) -> float:
        """片段时长"""
        return self.end - self.start

    @property
    def source_duration(self) -> float:
        """源素材使用的时长"""
        return (self.trim_end - self.trim_start) / self.speed

    def to_dict(self) -> Dict:
        return {
            "id": self.id,
            "material_id": self.material_id,
            "material_name": self.material_name,
            "start": self.start,
            "end": self.end,
            "trim_start": self.trim_start,
            "trim_end": self.trim_end,
            "speed": self.speed,
            "volume": self.volume,
            "effects": self.effects
        }

    @classmethod
    def from_dict(cls, data: Dict) -> "TimelineClip":
        return cls(
            id=data["id"],
            material_id=data["material_id"],
            material_name=data["material_name"],
            start=data["start"],
            end=data["end"],
            trim_start=data.get("trim_start", 0.0),
            trim_end=data.get("trim_end", 0.0),
            speed=data.get("speed", 1.0),
            volume=data.get("volume", 1.0),
            effects=data.get("effects", [])
        )


@dataclass
class Track:
    """轨道"""
    id: str
    type: str  # video, audio, subtitle
    name: str
    clips: List[TimelineClip] = field(default_factory=list)
    muted: bool = False
    locked: bool = False

    def add_clip(self, clip: TimelineClip):
        """添加片段"""
        self.clips.append(clip)
        self._sort_clips()

    def remove_clip(self, clip_id: str):
        """移除片段"""
        self.clips = [c for c in self.clips if c.id != clip_id]

    def _sort_clips(self):
        """按开始时间排序"""
        self.clips.sort(key=lambda c: c.start)

    def to_dict(self) -> Dict:
        return {
            "id": self.id,
            "type": self.type,
            "name": self.name,
            "clips": [c.to_dict() for c in self.clips],
            "muted": self.muted,
            "locked": self.locked
        }

    @classmethod
    def from_dict(cls, data: Dict) -> "Track":
        return cls(
            id=data["id"],
            type=data["type"],
            name=data["name"],
            clips=[TimelineClip.from_dict(c) for c in data.get("clips", [])],
            muted=data.get("muted", False),
            locked=data.get("locked", False)
        )


@dataclass
class Transition:
    """转场效果"""
    id: str
    type: str  # cut, dissolve, wipe, fade
    position: float  # 转场发生的时间点
    duration: float = 0.5
    params: Dict = field(default_factory=dict)

    def to_dict(self) -> Dict:
        return {
            "id": self.id,
            "type": self.type,
            "position": self.position,
            "duration": self.duration,
            "params": self.params
        }

    @classmethod
    def from_dict(cls, data: Dict) -> "Transition":
        return cls(
            id=data["id"],
            type=data["type"],
            position=data["position"],
            duration=data.get("duration", 0.5),
            params=data.get("params", {})
        )


@dataclass
class Timeline:
    """时间线"""
    id: str
    project_id: int
    video_track: Track = None
    audio_track: Track = None
    subtitle_track: Track = None
    transitions: List[Transition] = field(default_factory=list)
    duration: float = 0.0
    created_at: datetime = field(default_factory=datetime.utcnow)
    updated_at: datetime = field(default_factory=datetime.utcnow)

    def __post_init__(self):
        if self.video_track is None:
            self.video_track = Track(
                id=f"{self.id}_video",
                type="video",
                name="视频轨道"
            )
        if self.audio_track is None:
            self.audio_track = Track(
                id=f"{self.id}_audio",
                type="audio",
                name="音频轨道"
            )
        if self.subtitle_track is None:
            self.subtitle_track = Track(
                id=f"{self.id}_subtitle",
                type="subtitle",
                name="字幕轨道"
            )

    def add_clip(self, clip: TimelineClip, track_type: str = "video"):
        """添加片段到指定轨道"""
        track = self._get_track(track_type)
        if track:
            track.add_clip(clip)
            self._update_duration()

    def _get_track(self, track_type: str) -> Optional[Track]:
        """获取指定类型的轨道"""
        if track_type == "video":
            return self.video_track
        elif track_type == "audio":
            return self.audio_track
        elif track_type == "subtitle":
            return self.subtitle_track
        return None

    def _update_duration(self):
        """更新总时长"""
        max_end = 0.0
        for track in [self.video_track, self.audio_track, self.subtitle_track]:
            if track and track.clips:
                track_max = max(c.end for c in track.clips)
                max_end = max(max_end, track_max)
        self.duration = max_end

    def add_transition(self, transition: Transition):
        """添加转场"""
        self.transitions.append(transition)
        self.transitions.sort(key=lambda t: t.position)

    def to_dict(self) -> Dict:
        return {
            "id": self.id,
            "project_id": self.project_id,
            "duration": self.duration,
            "video_track": self.video_track.to_dict() if self.video_track else None,
            "audio_track": self.audio_track.to_dict() if self.audio_track else None,
            "subtitle_track": self.subtitle_track.to_dict() if self.subtitle_track else None,
            "transitions": [t.to_dict() for t in self.transitions],
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

    def to_json(self) -> str:
        return json.dumps(self.to_dict(), ensure_ascii=False)

    @classmethod
    def from_dict(cls, data: Dict) -> "Timeline":
        timeline = cls(
            id=data["id"],
            project_id=data["project_id"],
            duration=data.get("duration", 0.0),
            transitions=[Transition.from_dict(t) for t in data.get("transitions", [])]
        )
        if data.get("video_track"):
            timeline.video_track = Track.from_dict(data["video_track"])
        if data.get("audio_track"):
            timeline.audio_track = Track.from_dict(data["audio_track"])
        if data.get("subtitle_track"):
            timeline.subtitle_track = Track.from_dict(data["subtitle_track"])
        return timeline


@dataclass
class ClipPlan:
    """剪辑方案"""
    project_id: int
    clips: List[Dict] = field(default_factory=list)
    transitions: List[Dict] = field(default_factory=list)
    audio: Dict = field(default_factory=dict)
    total_duration: float = 0.0
    style: str = "动感"
    created_at: datetime = field(default_factory=datetime.utcnow)

    def to_dict(self) -> Dict:
        return {
            "project_id": self.project_id,
            "clips": self.clips,
            "transitions": self.transitions,
            "audio": self.audio,
            "total_duration": self.total_duration,
            "style": self.style,
            "created_at": self.created_at.isoformat() if self.created_at else None
        }

    def to_json(self) -> str:
        return json.dumps(self.to_dict(), ensure_ascii=False)

    @classmethod
    def from_dict(cls, data: Dict) -> "ClipPlan":
        return cls(
            project_id=data["project_id"],
            clips=data.get("clips", []),
            transitions=data.get("transitions", []),
            audio=data.get("audio", {}),
            total_duration=data.get("total_duration", 0.0),
            style=data.get("style", "动感")
        )


def generate_id() -> str:
    """生成唯一 ID"""
    import uuid
    return str(uuid.uuid4())[:8]

File: synthetix-vue/src/store/modules/timeline.js
import { defineStore } from 'pinia'
import { timelineApi } from '@/api/modules'

export const useTimelineStore = defineStore('timeline', {
  state: () => ({
    tracks: [
      { id: 'v1', type: 'video', name: '视频', clips: [], muted: false, locked: false },
      { id: 'a1', type: 'audio', name: '音频 1', clips: [], muted: false, locked: false },
      { id: 'a2', type: 'audio', name: '音频 2', clips: [], muted: false, locked: false },
      { id: 's1', type: 'subtitle', name: '字幕', clips: [], muted: false, locked: false },
      { id: 'm1', type: 'marker', name: '标记', clips: [], muted: false, locked: false },
    ],
    duration: 0,
    zoom: 1.0,       // 1.0 = 1 pixel per second
    scrollX: 0,
    playheadPosition: 0,
    snapEnabled: true,
    selectedClipId: null,
    // AI 建议覆盖
    suggestedClips: [],
  }),

  getters: {
    totalWidth: (state) => Math.max(state.duration, 60) * state.zoom,
    playheadPercent: (state) => state.duration > 0 ? (state.playheadPosition / state.duration) * 100 : 0,
    videoTrack: (state) => state.tracks.find(t => t.type === 'video'),
    audioTracks: (state) => state.tracks.filter(t => t.type === 'audio'),
    subtitleTrack: (state) => state.tracks.find(t => t.type === 'subtitle'),
    markerTrack: (state) => state.tracks.find(t => t.type === 'marker'),
  },

  actions: {
    loadFromTimelineData(data) {
      if (!data) return
      // 从后端 timeline_data JSON 加载
      const trackTypes = ['video_track', 'audio_track', 'subtitle_track']
      const typeMap = { video_track: 'video', audio_track: 'audio', subtitle_track: 'subtitle' }

      for (const key of trackTypes) {
        const trackData = data[key]
        if (!trackData) continue
        const type = typeMap[key]
        const track = this.tracks.find(t => t.type === type)
        if (track && trackData.clips) {
          track.clips = trackData.clips.map(c => ({
            id: c.id,
            materialId: c.material_id,
            materialName: c.material_name,
            start: c.start,
            end: c.end,
            trimStart: c.trim_start || 0,
            trimEnd: c.trim_end || 0,
            speed: c.speed || 1,
            volume: c.volume || 1,
            marginBefore: c.margin_before || 0,
            marginAfter: c.margin_after || 0,
          }))
        }
      }

      // 从剪辑方案加载到视频轨道
      this._updateDuration()
    },

    loadFromPlanData(planData) {
      if (!planData || !planData.clips) return
      const videoTrack = this.tracks.find(t => t.id === 'v1')
      if (!videoTrack) return

      let currentTime = 0
      videoTrack.clips = planData.clips.map((clip, i) => {
        const startSec = parseTimeToSeconds(clip.start_time)
        const endSec = parseTimeToSeconds(clip.end_time)
        const clipStart = currentTime
        const clipEnd = clipStart + (endSec - startSec)
        currentTime = clipEnd
        return {
          id: `plan_${i}`,
          materialId: clip.material_id,
          materialName: clip.material_name || clip.purpose || `片段 ${i + 1}`,
          start: clipStart,
          end: clipEnd,
          trimStart: startSec,
          trimEnd: endSec,
          speed: 1,
          volume: 1,
          marginBefore: 0,
          marginAfter: 0,
          purpose: clip.purpose,
        }
      })
      this._updateDuration()
    },

    _updateDuration() {
      let maxEnd = 0
      for (const track of this.tracks) {
        for (const clip of track.clips) {
          maxEnd = Math.max(maxEnd, clip.end)
        }
      }
      this.duration = maxEnd
    },

    setZoom(zoom) {
      this.zoom = Math.max(0.1, Math.min(100, zoom))
    },

    zoomIn() { this.setZoom(this.zoom * 1.3) },
    zoomOut() { this.setZoom(this.zoom / 1.3) },

    setScrollX(x) {
      this.scrollX = Math.max(0, x)
    },

    setPlayheadPosition(position) {
      this.playheadPosition = Math.max(0, Math.min(this.duration, position))
    },

    selectClip(clipId) {
      this.selectedClipId = clipId
    },

    deselectClip() {
      this.selectedClipId = null
    },

    // 添加 AI 建议的片段（显示为虚线）
    addSuggestedClip(trackType, clip) {
      this.suggestedClips.push({ ...clip, trackType })
    },

    clearSuggestedClips() {
      this.suggestedClips = []
    },

    // 保存到后端
    async save(projectId) {
      if (!projectId) return
      try {
        const timelineData = {
          video_track: this.tracks.find(t => t.id === 'v1')?.toJSON() || null,
          audio_track: this.tracks.find(t => t.id === 'a1')?.toJSON() || null,
          subtitle_track: this.tracks.find(t => t.id === 's1')?.toJSON() || null,
        }
        await timelineApi.save(projectId, { timeline_data: timelineData })
      } catch (error) {
        console.error('保存时间线失败:', error)
      }
    },

    reset() {
      this.tracks.forEach(t => { t.clips = [] })
      this.duration = 0
      this.zoom = 1.0
      this.scrollX = 0
      this.playheadPosition = 0
      this.selectedClipId = null
      this.suggestedClips = []
    },
  },
})

// 工具函数
function parseTimeToSeconds(timeStr) {
  if (typeof timeStr === 'number') return timeStr
  if (!timeStr) return 0
  const parts = String(timeStr).split(':')
  if (parts.length === 3) {
    return parseInt(parts[0]) * 3600 + parseInt(parts[1]) * 60 + parseFloat(parts[2])
  }
  if (parts.length === 2) {
    return parseInt(parts[0]) * 60 + parseFloat(parts[1])
  }
  return parseFloat(timeStr) || 0
}
